sdk/identity/azure-identity/src/private/token_cache_internals.hpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Token cache internals and test hooks.
+ */
+
+#pragma once
+
+#include "token_cache.hpp"
+
+#include <azure/core/credentials/credentials.hpp>
+
+#include <chrono>
+#include <functional>
+#include <map>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+namespace Azure { namespace Identity { namespace _detail {
+  /**
+   * @brief Implements internal aspects of token cache and provides test hooks.
+   *
+   */
+  class TokenCache::Internals final {
+  private:
+    Internals() = delete;
+    ~Internals() = delete;
+
+  public:
+    /**
+     * @brief Represents a unique set of characteristics that are used to distinguish between cache
+     * entries.
+     *
+     */
+    struct CacheKey
+    {
+      std::string TenantId; ///< Tenant ID.
+      std::string ClientId; ///< Client ID.
+      std::string AuthorityHost; ///< Authority Host.
+      std::string Scopes; ///< Authentication Scopes as a single string.
+
+      bool operator<(TokenCache::Internals::CacheKey const& other) const;
+    };
+
+    /**
+     * @brief The cache itself.
+     *
+     */
+    static std::map<CacheKey, Core::Credentials::AccessToken> Cache;
+
+    /**
+     * @brief Vector of expiration for the cache entries, from earliest to the latest.
+     *
+     */
+    static std::vector<std::pair<decltype(Core::Credentials::AccessToken::ExpiresOn), CacheKey>>
+        Expirations;
+
+    /**
+     * @brief Cache and Expirations are kept in sync. The mutex is for accessing both of them.
+     *
+     */
+    static std::shared_timed_mutex Mutex;
+
+#if defined(TESTING_BUILD)
+    /**
+     * @brief A function that gets called before write lock is acquired.
+     *
+     */
+    static std::function<void()> OnBeforeWriteLock;
+#endif
```

Comment:
This scares me even more. We should not make the code we test and the code we ship too different.

Reply:
While agreeing on the principle, I respectfully disagree in this specific case - realistically this is the least of all evils. There is not a better way to achieve what we want, and if used carefully as we do, it is harmless. Even if half of the things would go wrong.

`OnBefore...` is a test hook. It is in `_detail` namespace in a header that is in `/src`, so not even available in includes. It is only available when the code is built with tests. SDK code should not be invoking it, only the test code. SDK code should not be setting the callback. If an SDK programmer tries to use it, a linker error will occur, the latest it can be notced is the nightly vcpkg build, but probably earlier.

Should you have built the SDK with test hooks, then you get one extra `nullptr` check. But it is between mutexes, so it is better to not do unnecessary things there. And it is not necessary to even check for null, in the ideal case. The worst thing that happens if the callback does not get invoked is that the test will fail / will fail to compile / link.

But it is not possible to test the multithreaded code with this quality of detail without these test hooks.